### PR TITLE
Kotlin: Refactor Backend configuration, resolve circular dependency

### PR DESCRIPTION
**Pull request by gizemb**
## What's new in this PR?

Jira issue: https://wearezeta.atlassian.net/browse/AN-6762

We need to be able to initialize Kotlin network layer with the selected backend configuration (prod, staging, qa or custom).

The implementation was done by Bradley on https://github.com/wireapp/wire-android/pull/2724 so I made some refactoring on this code to solve circular dependency between backendModule and networkModule.

I also refactored the code a little bit so that
- BackendRepository is responsible from all data operations regarding backend configuration
 (ex: Network layer asks BackendRepository what's the base url)
- Backend configuration is exposed as BackendItem, whether pre-defined (as prod, qa, staging), or custom.

### Testing

Unit tests are added/updated. 
Also manually tested with scenarios: (Kotlin on/off) x (pre-defined/custom backend)

#### APK
[Download build #1778](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1778/artifact/build/artifact/wire-dev-PR2749-1778.apk)

### CODE REVIEW

File: app/src/test/kotlin/com/waz/zclient/core/backend/datasources/local/BackendLocalDataSourceTest.kt [hunk 1]
```diff
@@ -16,63 +18,90 @@ class BackendLocalDataSourceTest : UnitTest() {
...
+        mockBackendPrefs(valid = true)
 
-        val response = backendPrefsDataSource.getCustomBackendConfig()
+        val response = backendPrefsDataSource.backendConfig()
 
-        assert(response.isRight)
+        response.fold({
+            assert(false) { "Expected a valid preference" }
+        }) {
+            assert(it == TEST_PREFERENCE) //TODO @Fernando : Does not verify with kluent: "it shouldBe TEST_PREFERENCE"
```
Comment (gizemb) [inline app/src/test/kotlin/com/waz/zclient/core/backend/datasources/local/BackendLocalDataSourceTest.kt:42]: @android10 here's an example where kluent does not work, but assert works.

Comment (android10) [reply app/src/test/kotlin/com/waz/zclient/core/backend/datasources/local/BackendLocalDataSourceTest.kt:42]: Thank you! I will investigate a bit why :heart:

Comment (android10) [reply app/src/test/kotlin/com/waz/zclient/core/backend/datasources/local/BackendLocalDataSourceTest.kt:42]: @gizemb can you provide a bit more details on what is going on when that "does not work"?
Does it freezes the execution? False negatives?

Comment (gizemb) [reply app/src/test/kotlin/com/waz/zclient/core/backend/datasources/local/BackendLocalDataSourceTest.kt:42]: It gives false negative

Comment (android10) [reply app/src/test/kotlin/com/waz/zclient/core/backend/datasources/local/BackendLocalDataSourceTest.kt:42]: Interesting! Thanks I will dig a bit deeper on this.

**Pull request merged**